### PR TITLE
Assert all the things!

### PR DESCRIPTION
**Pull request by NicolasDorier**
if pindexPrev was null, we would get a problem at `pindexPrev->GetMedianTimePast()`

### CODE REVIEW

**Comment by dcousens** [comment]
utACK 8a87470

**Comment by luke-jr** [comment]
This means we're not enforcing it on the genesis block? Doesn't look like a practical problem today, but feels like a booby-trap. :/

**Comment by NicolasDorier** [comment]
@luke-jr yes, there is two calls to it, one check explicitely it is not the genesis

https://github.com/NicolasDorier/bitcoin/blob/8a87470ebbd289b502e7aec5476368dcfc3c18bb/src/main.cpp#L3655

And the other is TestBlockValidity, which test a block which got mined, so obviously not the genesis.

https://github.com/NicolasDorier/bitcoin/blob/8a87470ebbd289b502e7aec5476368dcfc3c18bb/src/main.cpp#L3776

**Comment by NicolasDorier** [comment]
Added [one more assert](https://github.com/bitcoin/bitcoin/pull/8665/commits/34a37bad84330126b97fbc31ca1c01a9ed1c5664) which make review more easy. This one bit me quite a few time, as pIndex sometimes refers to the block being evaluated, and sometimes to the previous one.

**Comment by dcousens** [comment]
utACK 34a37ba

**Comment by NicolasDorier** [comment]
mmh failure interesting with last commit, investigating it.

**Comment by NicolasDorier** [comment]
yuk, the problem was that `pindex->phashBlock == NULL` is used as a flag to know it is called by CreateNewBlock...

**Comment by morcos** [comment]
I don't think we should tag PR's like this with "Trivial"
Perhaps I'm misremembering, but I thought "Trivial" was for things that only changed comments or strings that couldn't possibly change behavior.  This PR takes at least a little bit of thinking about.

I don't have much of an opinion on the actual changes.

**Comment by sipa** [comment]
I agree, this is not trivial. Trivial is for comment and documentation
changes. Not just "this does not change behaviour", but "this so obviously
does not change behaviour that it does not need review".

File: src/pow.cpp [hunk 1]
```diff
@@ -12,12 +12,9 @@
 
 unsigned int GetNextWorkRequired(const CBlockIndex* pindexLast, const CBlockHeader *pblock, const Consensus::Params& params)
 {
+    assert(pindexLast != NULL);
     unsigned int nProofOfWorkLimit = UintToArith256(params.powLimit).GetCompact();
 
-    // Genesis block
```
Comment (laanwj) [inline src/pow.cpp]: This case (Genesis block) is never actually necessary?

Comment (dcousens) [inline src/pow.cpp]: `GetNextWorkRequired` is never called with `pindexLast == NULL`

**Comment by NicolasDorier** [comment]
Actually I thought that we were in the "this so obviously does not change behaviour that it does not need review" condition, until I saw that phashBlock could be NULL...

If it is not obvious then maybe this PR is not worth the review time, I can remove some commits if needed to make it trivial again.

**Comment by dcousens** [comment]
I think the PR is still worth it for the simple fact it adds more information/context to the code.
However,  perhaps as discussed on IRC,  it might be better to just remove the assertions and de-reference the pointers at the call sites by using a straight C++ reference.

**Comment by laanwj** [comment]
> If it is not obvious then maybe this PR is not worth the review time, I can remove some commits if needed to make it trivial again.

I think these changes are okay, let's just not call it trivial. As @morcos says the trivial denomination is for puls that don't change code, e.g. documentation typos, string typos, comment changes. Asserts generate code and can in principle create bad situations like DoS attacks if they're remote-triggerable so need to be carefully reviewed.
(which doesn't mean that changes to assertions aren't worth review time)

**Comment by MarcoFalke** [comment]
utACK dc64a16af92bf459dde32c32cea11b266db54f04

**Comment by MarcoFalke** [comment]
Needs rebase.

**Comment by NicolasDorier** [comment]
rebased

File: src/validation.cpp [hunk 2]
```diff
@@ -1739,7 +1739,10 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
                   CCoinsViewCache& view, const CChainParams& chainparams, bool fJustCheck)
 {
     AssertLockHeld(cs_main);
-
+    assert(pindex);
+    // pindex->phashBlock can be null if called by CreateNewBlock
```
Comment (TheBlueMatt) [inline src/validation.cpp]: (or, generally, by TestBlockValidity)

**Review by TheBlueMatt — COMMENTED** [review]
utACK 61b84a0142cd736a2b819b771fc843703ab8ae65 (+/- comment on comment)

**Comment by NicolasDorier** [comment]
addressed TheBlueMatt comment

**Comment by NicolasDorier** [comment]
Anything missing ?

**Comment by NicolasDorier** [comment]
not sure what to do with this PR. I think it increases readability... but it has been open since september. Do I keep open ?

**Comment by laanwj** [comment]
Yeah let's merge it. Asserting out is better than crashing with a NULL pointer dereference.

**Pull request merged**